### PR TITLE
[analyzer][NFC] Add some docs for LazyCompoundValue

### PR DESCRIPTION
**Pull request by Szelethus**
Yes, I basically copy-pasted some posts from discord and Artem's book, but these make for a rather decent docs.

### CODE REVIEW

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h [hunk 1]
```diff
@@ -346,6 +352,39 @@ class CompoundVal : public NonLoc {
...
+///
+/// However, there is another compound value used in the analyzer, which appears
+/// much more often during analysis, which is nonloc::LazyCompoundVal. This
+/// value is an r-value that represents a snapshot of any structure "as a whole"
+/// at a given moment during the analysis. Such value is already quite far from
+/// being re- ferred to as "concrete", as many fields inside it would be unknown
+/// or symbolic. nonloc::LazyCompoundVal operates by storing two things:
+///   * a reference to the TypedValueRegion being snapshotted (yes, it is always
+///     typed), and also
+///   * a copy of the whole Store object, obtained from the ProgramState in
```
Comment (NagyDonat) [inline clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:368]: Is it truly a _copy_ of the store object? This seems to contradict the "is a very cheap operation" line in the next paragraph...?

Or is it a _copied store object_ that _shares the same bulk allocations_ with the original one (but differs in minor details?)

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h [hunk 2]
```diff
@@ -346,6 +352,39 @@ class CompoundVal : public NonLoc {
...
+/// The simplest example of a concrete compound value is nonloc::CompoundVal,
+/// which represents a concrete r-value of an initializer-list or a string.
+/// Internally, it contains an llvm::ImmutableList of SVal's stored inside the
+/// literal.
+///
+/// However, there is another compound value used in the analyzer, which appears
+/// much more often during analysis, which is nonloc::LazyCompoundVal. This
+/// value is an r-value that represents a snapshot of any structure "as a whole"
+/// at a given moment during the analysis. Such value is already quite far from
+/// being re- ferred to as "concrete", as many fields inside it would be unknown
```
Comment (NagyDonat) [inline clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:364]:
```suggestion
/// being referred to as "concrete", as many fields inside it would be unknown
```

**Review by NagyDonat — COMMENTED** [review]
I'm really happy that you decided to document these :smile:

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h (hunk 1 above)
Comment (NagyDonat) [reply clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:368]: Perhaps we should mention that `Store` is just a typedef for `const void *`...

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h [hunk 3]
```diff
@@ -326,6 +326,12 @@ class LocAsInteger : public NonLoc {
   static bool classof(SVal V) { return V.getKind() == LocAsIntegerKind; }
 };
 
+/// The simplest example of a concrete compound value is nonloc::CompoundVal,
+/// which represents a concrete r-value of an initializer-list or a string.
+/// Internally, it contains an llvm::ImmutableList of SVal's stored inside the
+/// literal.
+///
+/// Source: https://github.com/haoNoQ/clang-analyzer-guide, v0.1, section 5.3.2.
```
Comment (steakhal) [inline clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:334]: If we want to phase out the guide, I don't think it's valuable to mention it here.
If you insist, let's mention this additional resource in the dev docs, but not directly here.

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h [hunk 4]
```diff
@@ -346,6 +352,39 @@ class CompoundVal : public NonLoc {
...
+///   * a copy of the whole Store object, obtained from the ProgramState in
+///     which it was created.
+///
+/// Essentially, nonloc::LazyCompoundVal is a performance optimization for the
+/// analyzer. Because Store is immutable, creating a nonloc::LazyCompoundVal is
+/// a very cheap operation. Note that the Store contains all region bindings in
+/// the program state, not only related to the region. Later, if necessary, such
+/// value can be unpacked -- eg. when it is assigned to another variable.
+///
+/// Source: https://github.com/haoNoQ/clang-analyzer-guide, v0.1, section 5.3.2.
```
Comment (steakhal) [inline clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:377]:
```suggestion
```

**Review by steakhal — COMMENTED** [review]
Makes sense. It's good to see some love for the docs.

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h (hunk 3 above)
Comment (haoNoQ) [reply clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:334]: Yeah please nuke this thank you ^.^

Sometimes we cite the academic papers that describe the specific approaches we've implemented (IPA is a good example) but this one's definitely inside out. The source code should be the primary source of truth about the source code.

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h (hunk 1 above)
Comment (haoNoQ) [reply clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:368]: > Is it truly a copy of the store object? This seems to contradict the "is a very cheap operation" line in the next paragraph...?

Yes! Congratulations, now you understand functional programming. Because the Store is an ImmutableMap, it is uniquely identified by a single pointer to the unique, carefully deduplicated instance of that map. The cost of a copy is the cost of a single 64-bit assignment. It's literally cheaper to copy the entire thing than to copy a portion of it, because a portion would be a different object, which we haven't created yet. It's like copy on write, except, the write isn't supported in the first place.

Of course the best part is that the "modification" is also cheap. Instead of modifying a map (and ruining everyone's "copy"), we just describe the modified map as essentially "the same as the other map except this one key has a different value". Which is also very quick to describe. Though, we still need to organize them into AVL trees for both fast lookup and fast deduplication. So it's not as straightforward as immutable lists which all naturally "grow on trees". But still, yes, all these operations are cheap, and this is how when you print the exploded graph it prints you every state, every store, behind every node. Because it doesn't cost anything extra to keep the copies around.

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h [hunk 5]
```diff
@@ -346,6 +352,39 @@ class CompoundVal : public NonLoc {
...
+///
+/// If you ever need to see if a LazyCompoundVal is fully or partially
+/// (un)initialized, you can iterBindings(). This is non-typechecked lookup
+/// (i.e., you cannot figure out which specific sub-region is initialized by the
+/// value you look at, you only get a byte offset). You can also improve
+/// iterBindings() to make it possible to restrict the iteration to a single
+/// cluster, because within the LazyCompoundVal’s Store only the cluster that
+/// corresponds to the LazyCompoundVal’s parent region is relevant.
```
Comment (haoNoQ) [inline clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:385]:
```suggestion
/// If you ever need inspect the contents of the LazyCompoundVal, you can use
/// StoreManager::iterBindings(). It'll iterate through all values in the Store,
/// but you're only interested in the ones that belong to LazyCompoundVal::getRegion();
/// other bindings are immaterial.
/// 
/// NOTE: LazyCompoundVal::getRegion() itself is also immaterial. It is only an
/// implementation detail. LazyCompoundVal represents only the rvalue, the data (known or unknown)
/// that *was* stored in that region *at some point in the past*. The region should not be used for
/// any purpose other than figuring out what part of the frozen Store you're interested in.
/// The value does not represent the *current* value of that region. Sometimes it may, but
/// this should not be relied upon. Instead, if you want to figure out what region it represents,
/// you typically need to see where you got it from in the first place.
/// The region is absolutely not analogous to the C++ "this" pointer.
/// It is also not a valid way to "materialize" the prvalue into a glvalue in C++, because the region
/// represents the *old* storage (sometimes very old), not the *future* storage.
```

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h [hunk 6]
```diff
@@ -346,6 +352,39 @@ class CompoundVal : public NonLoc {
   static bool classof(SVal V) { return V.getKind() == CompoundValKind; }
 };
 
+/// The simplest example of a concrete compound value is nonloc::CompoundVal,
+/// which represents a concrete r-value of an initializer-list or a string.
+/// Internally, it contains an llvm::ImmutableList of SVal's stored inside the
+/// literal.
+///
+/// However, there is another compound value used in the analyzer, which appears
+/// much more often during analysis, which is nonloc::LazyCompoundVal. This
```
Comment (haoNoQ) [inline clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:361]:
```suggestion
/// While nonloc::CompoundVal covers a few simple use cases, the nonloc::LazyCompoundVal
/// is a more performant and flexible way to represent an rvalue of record type,
/// so it shows up much more frequently during analysis. This
```

**Review by haoNoQ — COMMENTED** [review]
Thanks a lot!! Some comments are very much overdue here.

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h (hunk 1 above)
Comment (NagyDonat) [reply clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:368]: I do understand functional programming (I had a phase when I learnt Haskell and for a few years I thought that it's the absolutely perfect language :star_struck: -- since then I'm more realistic but I still like it), I was just confused by the fact that this line emphasizes "a copy of the whole Store object" as opposed to the "reference to" in the previous line.

In a functional language the word "copy" is practically non-existent (we speak about values, not objects and their identity, so there is no reason to say that this is _a copy of_ e.g. that list -- we simply say that "this is that list"), while in C++ saying "copy" instead of "move" or "pointer/reference" is an important synonym of "we allocate lots of new memory", so I'd still prefer tweaking this line.

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h [hunk 7]
```diff
@@ -346,6 +352,39 @@ class CompoundVal : public NonLoc {
...
+/// value is an r-value that represents a snapshot of any structure "as a whole"
+/// at a given moment during the analysis. Such value is already quite far from
+/// being re- ferred to as "concrete", as many fields inside it would be unknown
+/// or symbolic. nonloc::LazyCompoundVal operates by storing two things:
+///   * a reference to the TypedValueRegion being snapshotted (yes, it is always
+///     typed), and also
+///   * a copy of the whole Store object, obtained from the ProgramState in
+///     which it was created.
```
Comment (NagyDonat) [inline clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:369]:
```suggestion
///  * a reference to the whole Store object, obtained from the ProgramState in
///    which the nonloc::LazyCompoundVal was created.
///
/// Note that the old ProgramState and its Store is kept alive during the
/// analysis because these are immutable functional data structures and each new
/// Store value is represented as "earlier Store" + "additional binding".
```

What would you think about this phrasing?

**Comment by Szelethus** [comment]
I'd also love to see some docs for `Loc` and `NonLoc`, because I recall some angry looks from years ago when I touched those on their own, but I can barely recall them :( As well as some docs for the various `ProgramState::getSVal` methods...

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h [hunk 8]
```diff
@@ -346,6 +350,36 @@ class CompoundVal : public NonLoc {
...
+/// an rvalue of record type, so it shows up much more frequently during
+/// analysis. This value is an r-value that represents a snapshot of any
+/// structure "as a whole" at a given moment during the analysis. Such value is
+/// already quite far from being referred to as "concrete", as many fields
+/// inside it would be unknown or symbolic. nonloc::LazyCompoundVal operates by
+/// storing two things:
+///   * a reference to the TypedValueRegion being snapshotted (yes, it is always
+///     typed), and also
+///  * a reference to the whole Store object, obtained from the ProgramState in
+///    which the nonloc::LazyCompoundVal was created.
```
Comment (steakhal) [inline clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:364]: These bulletpoints are indented by different levels.

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h [hunk 9]
```diff
@@ -346,6 +350,36 @@ class CompoundVal : public NonLoc {
...
+/// analysis because these are immutable functional data structures and each new
+/// Store value is represented as "earlier Store" + "additional binding".
+///
+/// Essentially, nonloc::LazyCompoundVal is a performance optimization for the
+/// analyzer. Because Store is immutable, creating a nonloc::LazyCompoundVal is
+/// a very cheap operation. Note that the Store contains all region bindings in
+/// the program state, not only related to the region. Later, if necessary, such
+/// value can be unpacked -- eg. when it is assigned to another variable.
+///
+/// If you ever need inspect the contents of the LazyCompoundVal, you can use
```
Comment (steakhal) [inline clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:376]:
```suggestion
/// If you ever need to inspect the contents of the LazyCompoundVal, you can use
```

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h [hunk 10]
```diff
@@ -363,6 +397,18 @@ class LazyCompoundVal : public NonLoc {
...
+  /// that *was* stored in that region *at some point in the past*. The region
+  /// should not be used for any purpose other than figuring out what part of
+  /// the frozen Store you're interested in. The value does not represent the
+  /// *current* value of that region. Sometimes it may, but this should not be
+  /// relied upon. Instead, if you want to figure out what region it represents,
+  /// you typically need to see where you got it from in the first place. The
+  /// region is absolutely not analogous to the C++ "this" pointer. It is also
+  /// not a valid way to "materialize" the prvalue into a glvalue in C++,
+  /// because the region represents the *old* storage (sometimes very old), not
+  /// the *future* storage.
```
Comment (steakhal) [inline clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:411]: To be fair, I don't really understand this.

To me, this region is the region that one would need to query from the associated `Store` to get the value this `LazyCompoundValue` actually boils down to.
So, a region, like every other region, only defines a memory location (an l-value), and there is not much one could do with such a region. But I find this true for any other `MemRegion` anywhere in the engine.

Probably I miss something, as I don't usually deal with LCVs, and it's been a while I last touched them.
So, to summarize, I don't really understand what `immaterial` means in this context or understand why we mention it for `LCV::getRegion()` instead of mention this for a `MemRegion`.

**Review by steakhal — APPROVED** [review]
I think it's already in a good shape. I found barely any problems with the wording.

File: clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h (hunk 10 above)
Comment (Szelethus) [reply clang/include/clang/StaticAnalyzer/Core/PathSensitive/SVals.h:411]: @haoNoQ I can't really counter this, can you help us out a bit? :)

**Review by steakhal — APPROVED** [review]

**Comment by Szelethus** [comment]
I decided to merge this. We are still miles ahead with these docs, and I don't wanna accidentally forget the PR.
